driver-core/src/main/com/mongodb/MongoClientSettings.java

```diff
@@ -0,0 +1,677 @@
+            block.apply(builder);
+            socketSettings = builder.build();
+            return this;
+        }
+
+        /**
+         * Applies the {@link SocketSettings.Builder} block and then sets the heartbeatSocketSettings.
+         *
+         * @param block the block to apply to the heartbeatSocketSettings.
+         * @return this
+         * @see MongoClientSettings#getHeartbeatSocketSettings()
+         */
+        public Builder applyToHeartbeatSocketSettings(final Block<SocketSettings.Builder> block) {
+            notNull("block", block);
+            SocketSettings.Builder builder = SocketSettings.builder(heartbeatSocketSettings);
+            block.apply(builder);
+            heartbeatSocketSettings = builder.build();
+            return this;
+        }
+
+        /**
+         * Applies the {@link ConnectionPoolSettings.Builder} block and then sets the connectionPoolSettings.
+         *
+         * @param block the block to apply to the ConnectionPoolSettings.
+         * @return this
+         * @see MongoClientSettings#getConnectionPoolSettings()
+         */
+        public Builder applyToConnectionPoolSettings(final Block<ConnectionPoolSettings.Builder> block) {
+            notNull("block", block);
+            ConnectionPoolSettings.Builder builder = ConnectionPoolSettings.builder(connectionPoolSettings);
+            block.apply(builder);
+            connectionPoolSettings = builder.build();
+            return this;
+        }
+
+        /**
+         * Applies the {@link ServerSettings.Builder} block and then sets the serverSettings.
+         *
+         * @param block the block to apply to the ServerSettings.
+         * @return this
+         * @see MongoClientSettings#getServerSettings()
+         */
+        public Builder applyToServerSettings(final Block<ServerSettings.Builder> block) {
+            notNull("block", block);
+            ServerSettings.Builder builder = ServerSettings.builder(serverSettings);
+            block.apply(builder);
+            serverSettings = builder.build();
+            return this;
+        }
+
+        /**
+         * Applies the {@link SslSettings.Builder} block and then sets the sslSettings.
+         *
+         * @param block the block to apply to the SslSettings.
+         * @return this
+         * @see MongoClientSettings#getSslSettings()
+         */
+        public Builder applyToSslSettings(final Block<SslSettings.Builder> block) {
+            notNull("block", block);
+            SslSettings.Builder builder = SslSettings.builder(sslSettings);
+            block.apply(builder);
+            sslSettings = builder.build();
+            return this;
+        }
+
+        /**
+         * Sets the cluster settings.
+         *
+         * @param clusterSettings the cluster settings
+         * @return this
+         * @see MongoClientSettings#getClusterSettings()
+         */
+        public Builder clusterSettings(final ClusterSettings clusterSettings) {
```

Comment:
Is there a design rationale for keeping these methods? I'm concerned that having both clusterSettings and applyClusterSettings (and so on for each settings) will be confusing to users. 

I see how they're useful in re-implementing the existing async MongoClientSettings but if that's the only reason to keep them then let's try to come up with something different that doesn't require them.  One option is to add an apply method to each of the builders, so in the async MongoClientSettings you could do:

       public Builder clusterSettings(final ClusterSettings clusterSettings) {
            wrappedBuilder.applyToClusterSettings(new Block<ClusterSettings.Builder>() {
                @Override
                public void apply(final ClusterSettings.Builder builder) {
                    builder.apply(clusterSettings);  // same as ClusterSettings.builder(clusterSettings) but acting on an existing builder
                }
            });
            return this;
        }

Reply:
The rationale was is to keep the sync and async designs.  With the aim of moving away from the "custom" async Settings in 4.0.

Making users, just change import path, is *much* easier than making users rewrite how they interact with the settings.

Reply:
Done.